File: dalesbred/src/test/java/org/dalesbred/DatabaseArraysTest.java
/*
 * Copyright (c) 2015 Evident Solutions Oy
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package org.dalesbred;

import org.dalesbred.datatype.SqlArray;
import org.jetbrains.annotations.NotNull;
import org.junit.Rule;
import org.junit.Test;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class DatabaseArraysTest {

    private final Database db = TestDatabaseProvider.createPostgreSQLDatabase();

    @Rule
    public final TransactionalTestsRule rule = new TransactionalTestsRule(db);

    @Test
    public void databaseArraysAsPrimitiveArrays() {
        assertThat(db.findUnique(int[].class, "values (cast ('{1,5,3}' as numeric array))"), is(new int[] { 1, 5, 3 }));
        assertThat(db.findUnique(long[].class, "values (cast ('{1,6,3}' as numeric array))"), is(new long[] { 1, 6, 3 }));
        assertThat(db.findUnique(short[].class, "values (cast ('{1,6,3}' as numeric array))"), is(new short[] { 1, 6, 3 }));
    }

    @Test
    public void databaseArraysAsWrapperArrays() {
        assertThat(db.findUnique(Integer[].class, "values (cast ('{1,5,3}' as numeric array))"), is(new Integer[] { 1, 5, 3 }));
        assertThat(db.findUnique(Long[].class, "values (cast ('{1,6,3}' as numeric array))"), is(new Long[] { 1L, 6L, 3L }));
        assertThat(db.findUnique(Short[].class, "values (cast ('{1,6,3}' as numeric array))"), is(new Short[]{1, 6, 3}));
    }

    @Test
    public void databaseArraysForBigNumbers() {
        assertThat(db.findUnique(BigInteger[].class, "values (cast ('{1,5,3}' as numeric array))"), is(new BigInteger[]{BigInteger.valueOf(1), BigInteger.valueOf(5L), BigInteger.valueOf(3L)}));
        assertThat(db.findUnique(BigDecimal[].class, "values (cast ('{1,5,3}' as numeric array))"), is(new BigDecimal[]{BigDecimal.valueOf(1), BigDecimal.valueOf(5L), BigDecimal.valueOf(3L)}));
    }

    @Test
    public void databaseArraysForStrings() {
        assertThat(db.findUnique(String[].class, "values (cast ('{foo,bar,baz}' as varchar array))"), is(new String[] { "foo", "bar", "baz" }));
    }

    @Test
    public void databaseArraysAsLists() {
        ListContainer object = db.findUnique(ListContainer.class, "select (cast ('{1,5,3}' as numeric array)) as intList, (cast ('{foo, bar, baz}' as varchar array)) as stringList");
        assertThat(object.intList, is(asList(1, 5, 3)));
        assertThat(object.stringList, is(asList("foo", "bar", "baz")));
    }

    @Test
    public void databaseArraysAsCollections() {
        CollectionContainer object = db.findUnique(CollectionContainer.class, "select (cast ('{1,5,3}' as numeric array)) as intCollection, (cast ('{foo, bar, baz}' as varchar array)) as stringCollection");
        assertThat(object.intCollection, is(asList(1, 5, 3)));
        assertThat(object.stringCollection, is(asList("foo", "bar", "baz")));
    }

    @Test
    public void databaseArraysAsSets() {
        SetContainer object = db.findUnique(SetContainer.class, "select (cast ('{1,5,3}' as numeric array)) as intSet, (cast ('{foo, bar, baz}' as varchar array)) as stringSet");
        assertThat(object.intSet, is(setOf(1, 5, 3)));
        assertThat(object.stringSet, is(setOf("foo", "bar", "baz")));
    }

    @Test
    public void bindArray() {
        db.update("drop table if exists array_test");
        db.update("create table array_test (string_array VARCHAR array)");

        db.update("insert into array_test (string_array) values (?)", SqlArray.varchars("foo", "bar"));

        assertThat(db.findUnique(String[].class, "select string_array from array_test"), is(new String[] { "foo", "bar" }));
    }

    @SuppressWarnings("unused")
    public static final class ListContainer {
        public List<Integer> intList;
        public List<String> stringList;
    }

    @SuppressWarnings("unused")
    public static final class SetContainer {
        public Set<Integer> intSet;
        public Set<String> stringSet;
    }

    @SuppressWarnings("unused")
    public static final class CollectionContainer {
        public Collection<Integer> intCollection;
        public Collection<String> stringCollection;
    }

    @NotNull
    @SafeVarargs
    private static <T> Set<T> setOf(T... items) {
        return new HashSet<>(asList(items));
    }
}

File: dalesbred/src/test/kotlin/org/dalesbred/DatabaseArraysTest.kt
package org.dalesbred

import org.dalesbred.datatype.SqlArray
import org.dalesbred.testutils.DatabaseProvider.POSTGRESQL
import org.dalesbred.testutils.DatabaseTest
import org.dalesbred.testutils.transactionalTest
import org.junit.jupiter.api.Assertions.assertArrayEquals
import java.math.BigDecimal
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals

@DatabaseTest(POSTGRESQL)
class DatabaseArraysTest(private val db: Database) {

    @Test
    fun `database arrays as primitive arrays`() = transactionalTest(db) {
        assertArrayEquals(intArrayOf(1, 5, 3), db.findUnique(IntArray::class.java, "values (cast ('{1,5,3}' as numeric array))"))
        assertArrayEquals(longArrayOf(1, 6, 3), db.findUnique(LongArray::class.java, "values (cast ('{1,6,3}' as numeric array))"))
        assertArrayEquals(shortArrayOf(1, 6, 3), db.findUnique(ShortArray::class.java, "values (cast ('{1,6,3}' as numeric array))"))
    }

    @Test
    fun `database arrays as wrapper arrays`() = transactionalTest(db) {
        assertArrayEquals(arrayOf(1, 5, 3), db.findUnique(Array<Int>::class.java, "values (cast ('{1,5,3}' as numeric array))"))
        assertArrayEquals(arrayOf(1L, 6L, 3L), db.findUnique(Array<Long>::class.java, "values (cast ('{1,6,3}' as numeric array))"))
        assertArrayEquals(arrayOf<Short>(1, 6, 3), db.findUnique(Array<Short>::class.java, "values (cast ('{1,6,3}' as numeric array))"))
    }

    @Test
    fun `database arrays for big numbers`() = transactionalTest(db) {
        assertArrayEquals(arrayOf(BigInteger.valueOf(1), BigInteger.valueOf(5L), BigInteger.valueOf(3L)), db.findUnique(Array<BigInteger>::class.java, "values (cast ('{1,5,3}' as numeric array))"))
        assertArrayEquals(arrayOf(BigDecimal.valueOf(1), BigDecimal.valueOf(5L), BigDecimal.valueOf(3L)), db.findUnique(Array<BigDecimal>::class.java, "values (cast ('{1,5,3}' as numeric array))"))
    }

    @Test
    fun `database arrays for strings`() = transactionalTest(db) {
        assertArrayEquals(arrayOf("foo", "bar", "baz"), db.findUnique(Array<String>::class.java, "values (cast ('{foo,bar,baz}' as varchar array))"))
    }

    @Test
    fun `database arrays as lists`() = transactionalTest(db) {
        val obj = db.findUnique(ListContainer::class.java, "select (cast ('{1,5,3}' as numeric array)) as intList, (cast ('{foo, bar, baz}' as varchar array)) as stringList")
        assertEquals(listOf(1, 5, 3), obj.intList)
        assertEquals(listOf("foo", "bar", "baz"), obj.stringList)
    }

    @Test
    fun `database arrays as collections`() = transactionalTest(db) {
        val obj = db.findUnique(CollectionContainer::class.java, "select (cast ('{1,5,3}' as numeric array)) as intCollection, (cast ('{foo, bar, baz}' as varchar array)) as stringCollection")
        assertEquals(listOf(1, 5, 3), obj.intCollection)
        assertEquals(listOf("foo", "bar", "baz"), obj.stringCollection)
    }

    @Test
    fun `database arrays as sets`() = transactionalTest(db) {
        val obj = db.findUnique(SetContainer::class.java, "select (cast ('{1,5,3}' as numeric array)) as intSet, (cast ('{foo, bar, baz}' as varchar array)) as stringSet")
        assertEquals(setOf(1, 5, 3), obj.intSet)
        assertEquals(setOf("foo", "bar", "baz"), obj.stringSet)
    }

    @Test
    fun `bind array`() = transactionalTest(db) {
        db.update("drop table if exists array_test")
        db.update("create table array_test (string_array VARCHAR array)")

        db.update("insert into array_test (string_array) values (?)", SqlArray.varchars("foo", "bar"))

        assertArrayEquals(arrayOf("foo", "bar"), db.findUnique(Array<String>::class.java, "select string_array from array_test"))
    }

    class ListContainer {
        lateinit var intList: List<Int>
        lateinit var stringList: List<String>
    }

    class SetContainer {
        lateinit var intSet: Set<Int>
        lateinit var stringSet: Set<String>
    }

    class CollectionContainer {
        lateinit var intCollection: Collection<Int>
        lateinit var stringCollection: Collection<String>
    }
}
